### PR TITLE
Early stopping on growing log

### PR DESCRIPTION
**Pull request by darksurfer**
Hello, 
I just made this pr to show a idea.  The early stopping was made to be applied on loss or decreasing values. 
I'm just proposing to letting it be applied on growing values like accuracy or val_acc. That being said  I don't know if this is a bad pattern or not.

BTW : congratulations guys ! Keras is awesome you do an excellent job !

### CODE REVIEW

**Comment by fchollet** [comment]
Agreed that it can be useful. Not a huge fan of the interface though (additional keyword argument `order=...`). Can we find something better? For starters, I would find something like `mode='max'` (stop at/close to max of function) and `mode='min'` to be less cognitively intensive. 

Wouldn't it be safe to assume that when monitoring an accuracy, we should stop when accuracy is no longer increasing, by default? We would only apply such a default if no order is explicitly specified.

**Comment by darksurfer** [comment]
So what about something like this : https://gist.github.com/darksurfer/e8e84c8d0ec9452ebba7#file-early-py-L10 , which use a max function for accuracy and let custom function being used if needed.

**Comment by fchollet** [comment]
This seems quite reasonable.

A name like `mode` would probably be more meaningful than `cmp`, though.

**Comment by fchollet** [comment]
Any interest in updating the PR? Otherwise I'll close it.

**Comment by darksurfer** [comment]
Yep Sorry i jus't didn't have time enough ! but i'm still interested .

File: keras/callbacks.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def on_epoch_end(self, epoch, logs={}):
 
 
 class EarlyStopping(Callback):
-    def __init__(self, monitor='val_loss', patience=0, verbose=0):
+    def __init__(self, monitor='val_loss', patience=0, verbose=0,mode=None):
```
Comment (fchollet) [inline keras/callbacks.py]: Space after `,`.

File: keras/callbacks.py [hunk 2]
```diff
@@ -211,13 +211,20 @@ def __init__(self, monitor='val_loss', patience=0, verbose=0):
...
+            else:
+                self.mode = min
 
     def on_epoch_end(self, epoch, logs={}):
         current = logs.get(self.monitor)
         if current is None:
             warnings.warn("Early stopping requires %s available!" % (self.monitor), RuntimeWarning)
 
-        if current < self.best:
+        if self.mode(current,self.best) == current:
```
Comment (fchollet) [inline keras/callbacks.py]: Space after `,`.

**Comment by fchollet** [comment]
You could use the following method to get a single commit: http://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by wuaalb** [comment]
Maybe you should do something like initialize `self.best = None` and check in the condition?

Without reading the actual implementation for me it isn't very clear what `mode` does or how it works.
A 'min' or 'max' mode (string), without automatic `None` option would be more intuitive for me (but less flexible and fool-proof I guess).

**Comment by darksurfer** [comment]
Wow, I try to rebase -i on a the few commit i made and this absolutely doesn't go as planed :/

**Comment by darksurfer** [comment]
Okay I made the early Stopping function more pep8 compliant ( thanks to pymode). 
If this code is acceptable for you I’ll make another clean  pull request  based on master.
What do you think ?

File: docs/sources/callbacks.md [hunk 3]
```diff
@@ -37,10 +37,11 @@ Save the model after every epoch. If `save_best_only=True`, the latest best mode
 
 
 ```python
-keras.callbacks.EarlyStopping(monitor='val_loss', patience=0, verbose=0)
+keras.callbacks.EarlyStopping(monitor='val_loss', patience=0, verbose=0,mode=None)
```
Comment (fchollet) [inline docs/sources/callbacks.md]: Space after comma

File: docs/sources/callbacks.md [hunk 4]
```diff
@@ -37,10 +37,11 @@ Save the model after every epoch. If `save_best_only=True`, the latest best mode
...
 ```
 
-Stop training after no improvement of the metric `monitor` is seen for `patience` epochs.
+Stop training after no improvement of the metric `monitor` is seen for `patience` epochs. 
+The `mode` parameter must be a function that compare two values and return the best one. Function like `min` or `max`. If no mode is specified `max` is used for metric that contain "acc"(i.e. "acc" or "val_acc")  and `min` otherwise.
```
Comment (fchollet) [inline docs/sources/callbacks.md]: Should specify "Python function" (as opposed to Theano)

File: keras/callbacks.py [hunk 5]
```diff
@@ -211,13 +212,20 @@ def __init__(self, monitor='val_loss', patience=0, verbose=0):
...
+        if mode:
+            self.mode = mode
+        else:
+            if "acc" in self.monitor:
+                self.mode = max
+            else:
+                self.mode = min
 
-    def on_epoch_end(self, epoch, logs={}):
+     def on_epoch_end(self, epoch, logs={}):
```
Comment (fchollet) [inline keras/callbacks.py]: What's going on with the indentation here?

**Comment by fchollet** [comment]
I think the API would be clearer if `mode` was a string from a given set `{'max', 'min', 'auto'}`. That would be more consistent with the rest of the codebase, and easier to understand IMO.

`auto` would be converted to `min` and `max` depending on the monitored metric.

File: keras/callbacks.py (hunk 5 above)
Comment (darksurfer) [inline keras/callbacks.py]: I think my editor screw me. Sorry i'm not used to those tools yet.

**Comment by fchollet** [comment]
Looking good, will merge this. Before I merge, please:

- use 4 space indentation (right now you are using 5 spaces)
- squash your commits into 1-2 commits with relevant messages, since this is a small change and shouldn't leave a ton of commits in the history.

Thanks!
